src/fastify-plugins/maps/index.js

```diff
@@ -20,57 +24,153 @@ export const plugin = fp(mapsPlugin, {
 /**
  * @typedef {object} MapsPluginOpts
  * @property {string} [prefix]
+ * @property {string} [defaultOnlineStyleUrl]
  */
 
 /** @type {import('fastify').FastifyPluginAsync<MapsPluginOpts>} */
 async function mapsPlugin(fastify, opts) {
   fastify.register(routes, {
     prefix: opts.prefix,
+    defaultOnlineStyleUrl: opts.defaultOnlineStyleUrl,
   })
 }
 
+const GetStyleJsonQueryStringSchema = T.Object({
+  key: T.Optional(T.String()),
+})
+
 /** @type {import('fastify').FastifyPluginAsync<MapsPluginOpts, import('fastify').RawServerDefault, import('@fastify/type-provider-typebox').TypeBoxTypeProvider>} */
-async function routes(fastify) {
-  fastify.get('/style.json', async (req, rep) => {
-    const serverAddress = await getFastifyServerAddress(req.server.server)
-
-    // 1. Attempt to get "default" local static map's style.json
-    {
-      const styleId = 'default'
-
-      const results = await Promise.all([
-        fastify.mapeoStaticMaps.getStyleJsonStats(styleId),
-        fastify.mapeoStaticMaps.getResolvedStyleJson(styleId, serverAddress),
-      ]).catch(() => {
-        fastify.log.warn('Cannot read default static map')
-        return null
-      })
-
-      if (results) {
+async function routes(fastify, opts) {
+  const { defaultOnlineStyleUrl = DEFAULT_MAPBOX_STYLE_URL } = opts
+
+  fastify.get(
+    '/style.json',
+    { schema: { querystring: GetStyleJsonQueryStringSchema } },
+    async (req, rep) => {
+      const serverAddress = await getFastifyServerAddress(req.server.server)
+
+      // 1. Attempt to get "default" local static map's style.json
+      {
+        const styleId = 'default'
+
+        const results = await Promise.all([
+          fastify.mapeoStaticMaps.getStyleJsonStats(styleId),
+          fastify.mapeoStaticMaps.getResolvedStyleJson(styleId, serverAddress),
+        ]).catch(() => {
+          fastify.log.warn('Cannot read default static map')
+          return null
+        })
+
+        if (results) {
+          const [stats, styleJson] = results
+          rep.headers(createStyleJsonResponseHeaders(stats.mtime))
+          return styleJson
+        }
+      }
+
+      // 2. Attempt to get a default style.json from online source
+      {
+        const { key } = req.query
+        const upstreamUrl = key
+          ? styleUrlWithApiKey(defaultOnlineStyleUrl, key)
+          : defaultOnlineStyleUrl
+
+        try {
+          const upstreamResponse = await fetch(upstreamUrl)
+
+          if (upstreamResponse.ok) {
+            // Set up headers to forward
+            for (const [name, value] of upstreamResponse.headers) {
```

Comment:
yeah that make sense. not sure about which headers are essential though (there are quite a few from these providers 😄 ). happy to change if we can narrow down the ones we want to actually forward

Reply:
@gmaclennan any thoughts?

Reply:
i updated this implementation in https://github.com/digidem/mapeo-core-next/pull/459/commits/ef1c0dd621781363835150d645c2dde83e8486b9 to use an allow-list approach and currently have it so that it's only forwarding headers related to caching.